### PR TITLE
support php8 changes

### PR DESCRIPTION
**Pull request by Shaked**
Following PHP 8 RFC https://wiki.php.net/rfc/string_to_number_comparison, you can no longer compare `0 == ''` and expect a `true` result. 

This PR fixes the following snippet which suppose to generate a SAS url: 

```
$sas_helper = new MicrosoftAzure\Storage\Blob\BlobSharedAccessSignatureHelper(
    $accountName, $accountKey);
$sas = $sas_helper->generateBlobServiceSharedAccessSignatureToken(
    MicrosoftAzure\Storage\Blob\Internal\BlobResources::RESOURCE_TYPE_BLOB,              # Resource name to generate the canonicalized resource. It can be Resources::RESOURCE_TYPE_BLOB or Resources::RESOURCE_TYPE_CONTAINER
    "{$containerName}/{$blobName}",                     # The name of the resource, including the path of the resource. It should be {container}/{blob}: for blobs.
    "r",                                        # Signed permissions.
    (new \DateTime())->modify('+10 minute'),    # Signed expiry
    (new \DateTime())->modify('-5 minute'),     # Signed start
    '',                                         # Signed IP, the range of IP addresses from which a request will be accepted, eg. "168.1.5.60-168.1.5.70"
    'https',                                    # Signed protocol, should always be https
);
echo "https://{$accountName}.blob.core.windows.net/{$containerName}/{$blobName}?{$sas}";
```

### CODE REVIEW

**Review by katmsft — APPROVED** [review]
LGTM.

**Comment by lukewawrzyniak** [comment]
Hi,
When will you merge this to master? thx

**Comment by yakmoose** [comment]
Any chance of this being merged soon?

**Comment by mdowton** [comment]
Also looking to have this fix merged pls ? I have same issue

**Comment by acato-dev** [comment]
Also looking to have this fix merged pls ? We're running into this same issue. Looking forward on an update on the matter :)
@katmsft

**Comment by sdmgeek** [comment]
Will this be merged soon?

**Comment by unknown2605** [comment]
Hope this PR will be merged soon

**Comment by miguelwicht** [comment]
Any info on if or when this will get merged?

**Comment by gintechsystems** [comment]
This really needs to get merged in order to use PHP 8, without any forking or local modifications. This is the only project holding us back from moving forward with PHP 8. Is there anyone from the Azure team that can see this PR? The fix is there we just need it merged. It works well too.

**Comment by spaze** [comment]
Since the release schedule is unpredictable (history repeating what was the same issue with PHP 7.4 support a year ago, when maintainers have promised promised [quarterly releases](https://github.com/Azure/azure-storage-php/issues/207#issuecomment-658531592) and it's now 6 months since the last release) I can only suggest local modifications using the [composer-patches](https://github.com/cweagans/composer-patches) plugin.

This is what we have done to have this fix (and others too) without needing to wait for a release:

1. Added extra config to `composer.json`:
```json
	"extra": {
		"patches-file": "patches/composer.patches.json"
	}
```

`patches/composer.patches.json` contains the following:
```json
{
	"patches": {
		"microsoft/azure-storage-common": {
			"Compare numbers to numbers": "patches/microsoft-azure-storage-common-php8-comparison-pr274.diff"
		},
		"microsoft/azure-storage-table": {
			"Sort callback must return int not bool": "patches/microsoft-azure-storage-table-php8-sort-callback-pr288.diff"
		}
	}
}
```

I've published the patch files (for this and other bugs) here https://gist.github.com/spaze/946b885dc668397bf0386eefeebf5ad1 feel free to use them.

**Comment by uobmed** [comment]
The fix by Spaze needs to be merged for PHP 8.

**Comment by ChristianVermeulen** [comment]
@katmsft Any chance we can bump this in prio a bit? This is blocking everyone from upgrading to php 8.0 :'(.

**Comment by johanrosenson** [comment]
This is an official microsoft repository right? 

Do we need to call Bill?

**Comment by gintechsystems** [comment]
It is awesome to see that this PR finally got merged. For composer sake, we need a new version tag. Please and thanks!

**Comment by spaze** [comment]
> It is awesome to see that this PR finally got merged. For composer sake, we need a new version tag. Please and thanks!

Don't hold your breath though :-) Composer releases are done from a different repository, https://github.com/Azure/azure-storage-common-php in this case. And there are some other things failing on PHP 8, like #287. This is sad, however it's not the first time, unfortunately.

**Comment by katmsft** [comment]
> > It is awesome to see that this PR finally got merged. For composer sake, we need a new version tag. Please and thanks!
> 
> Don't hold your breath though :-) Composer releases are done from a different repository, https://github.com/Azure/azure-storage-common-php in this case. And there are some other things failing on PHP 8, like #287. This is sad, however it's not the first time, unfortunately.

Could you or anyone else provide a full PR list to enable PHP 8 support, please?

**Comment by ChristianVermeulen** [comment]
@katmsft @spaze As far as i can see, this repo has a treesplit into the dedicated package repo's and every package has it's own release. Issue #287 is about getting the `table` package ready for PHP 8. This one (common) is just the base on which the others depend.

So in my opinion, it would be perfectly possible to already create a `v1.5.3-common` or even `v.1.6.0-common` tag so this new version is released already. This way, the other packages can already make use of this while the upgrade of the `table` package is pending.

**Comment by spaze** [comment]
> Could you or anyone else provide a full PR list to enable PHP 8 support, please?

Hi @katmsft I believe these two (#274 & #287) are all that's needed. We've been running on PHP 8 in dev for some 6 months and few weeks in production. However we're using only table, queue, blob (& common) packages, not file.

It would be cool to release other PRs too. Need help with maintaining? :-)

**Comment by katmsft** [comment]
> > Could you or anyone else provide a full PR list to enable PHP 8 support, please?
> 
> Hi @katmsft I believe these two (#274 & #287) are all that's needed. We've been running on PHP 8 in dev for some 6 months and few weeks in production. However we're using only table, queue, blob (& common) packages, not file.
> 
> It would be cool to release other PRs too. Need help with maintaining? :-)

Thanks for the kind offer but due to the sensitivity of the SDK (it deals with customer data) we cannot ask community to maintain this repo.

I have done some local test, it seems that some dependencies for engsys (phpdoc, phpunit) is not yet compatible with PHP 8 so that next version's PHP 8 support won't be comprehensive.

With this said we will release a new release soon but only take community PRs that resolve known PHP 8 incompatibilities.

**Comment by ChristianVermeulen** [comment]
@katmsft As far as I can tell, both phpdoc and phpunit are fully compatible with PHP8. Even phpunit 8 is already compatible. Could you elaborate a bit more on the problems with these dependencies perhaps?

**Comment by spaze** [comment]
> Thanks for the kind offer but due to the sensitivity of the SDK (it deals with customer data) we cannot ask community to maintain this repo.

@katmsft  Want to hire me then to help with maintenance to make it official? :-) I already worked for Microsoft once if that would make things easier...

**Pull request merged**
